src/uart.rs

```diff
@@ -687,11 +700,82 @@ impl<'a> UartRx<'a, Async> {
 
         let rx_dma = dma::Dma::reserve_channel(rx_dma);
 
-        Ok(Self::new_inner::<T>(Some(rx_dma)))
+        Ok(Self::new_inner::<T>(Some(rx_dma), None))
+    }
+
+    /// create a new DMA enabled UART which can only receive data, but uses a buffer to avoid FIFO overflow
+    /// Note: requires time-driver due to hardware constraint requiring a polled interface (no UART Idle bus indicator)
+    ///       alternative approaches are possible, this was done as it maintains most similarity between Buffered and
+    ///       Unbuffered read interfaces
+    #[cfg(feature = "time")]
+    pub fn new_async_with_buffer<T: Instance>(
+        _inner: impl Peripheral<P = T> + 'a,
+        rx: impl Peripheral<P = impl RxPin<T>> + 'a,
+        _irq: impl interrupt::typelevel::Binding<T::Interrupt, InterruptHandler<T>> + 'a,
+        rx_dma: impl Peripheral<P = impl RxDma<T>> + 'a,
+        config: Config,
+        buffer: &'static mut [u8],
+        polling_rate_us: u64,
+    ) -> Result<Self> {
+        assert!(buffer.len() <= 1024);
+
+        into_ref!(_inner);
+        into_ref!(rx);
+        rx.as_rx();
+
+        let mut _rx = rx.map_into();
+        Uart::<Async>::init::<T>(None, Some(_rx.reborrow()), None, None, config)?;
+
+        T::Interrupt::unpend();
+        unsafe { T::Interrupt::enable() };
+
+        let rx_dma = dma::Dma::reserve_channel(rx_dma);
+        T::info().regs.fifocfg().modify(|_, w| w.dmarx().enabled());
+        // immediately configure and enable channel for circular buffered reception
+        rx_dma.configure_channel(
+            dma::transfer::Direction::PeripheralToMemory,
+            T::info().regs.fiford().as_ptr() as *const u8 as *const u32,
+            buffer as *mut [u8] as *mut u32,
+            buffer.len(),
+            dma::transfer::TransferOptions {
+                width: dma::transfer::Width::Bit8,
+                priority: dma::transfer::Priority::Priority0,
+                is_continuous: true,
+                is_sw_trig: true,
+            },
+        );
+        rx_dma.enable_channel();
+        rx_dma.trigger_channel();
+
+        Ok(Self::new_inner::<T>(
+            Some(rx_dma),
+            Some(BufferConfig {
+                buffer,
+                write_index: 0,
+                read_index: 0,
+                polling_rate: polling_rate_us,
+            }),
+        ))
     }
 
     /// Read from UART RX asynchronously.
     pub async fn read(&mut self, buf: &mut [u8]) -> Result<()> {
+        #[cfg(feature = "time")]
+        {
+            if self._buffer_config.is_some() {
+                self.read_buffered(buf).await
+            } else {
+                self.read_unbuffered(buf).await
+            }
+        }
+
+        #[cfg(not(feature = "time"))]
+        {
+            self.read_unbuffered(buf).await
+        }
```

Comment:
If you had your separate `buffered.rs` submodule, this wouldn't be necessary. You would have your isolated `read()` implementation.